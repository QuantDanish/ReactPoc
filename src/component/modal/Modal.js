import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ModalTemplate from './ModalTemplate';
import ModalBody from './ModalBody';

let propTypes = {
  isOpen: PropTypes.bool.isRequired,
  closeHandler: PropTypes.func.isRequired,
  delay: PropTypes.number,
  className: PropTypes.string,
  modalStyle: PropTypes.object,
  backdropStyle: PropTypes.object,
  header: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.element
  ]),
  content: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.element
  ]),
  footer: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.element
  ])
}

let defaultProps = {
  isOpen: false,
  delay: 200,
  className: "",
  modalStyle: {},
  backdropStyle: {}
}

class Modal extends Component {
  constructor(props){
    super();
    this.state = {
        show: false,
        animation: false
    };
  }
  
  setShow(value){
    this.setState({
      animation: this.state.animation,
      show: value
    });
  }

  setAnimation(value){
    this.setState({
      animation: value,
      show: this.state.show});
  }

  renderModal () {
    this.setShow(true);
    setTimeout(() => {
      this.setAnimation(true);
    }, this.props.delay);
  };
  
  closeModal() {
    console.log("close  modal ::", this.state ,this.props)
    this.setAnimation(false);
    setTimeout(() => {
      this.setShow(false);
    }, this.props.delay);
  }
  
  shouldComponentUpdate(nextProps, nextState) {
    if(!this.props.isOpen && nextProps.isOpen){
      // when user click on button to open a modal.
      this.renderModal();
      return false;
    }    
    else if(this.props.isOpen && !nextProps.isOpen){
      // when use click on close button.
      this.closeModal();
      return false;
    }
    
    return true;
    
  }

  render() {
    const body = this.props.children || <ModalBody 
                                header={this.props.header} 
                                footer={this.props.footer} 
                                content={this.props.content} 
                                hideCallback={this.props.closeHandler} />;
    return (
      <ModalTemplate
        show={this.state.show}
        hideCallback= {this.props.closeHandler}
        startAnimation= {this.state.animation}
        body={body}
        className={this.props.className}
        modalStyle={this.props.modalStyle}
        backdropStyle={this.props.backdropStyle}
      ></ModalTemplate>
    )
  }

  componentDidMount(){
    if(this.props.isOpen) {
      this.renderModal();
    }
  }
}

Modal.propTypes = propTypes;
Modal.defaultProps = defaultProps;

export default Modal
