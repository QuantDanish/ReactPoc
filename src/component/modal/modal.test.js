import React from 'react';
import {configure, shallow, mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import Modal from './Modal';



configure({adapter: new Adapter()});

/**
|--------------------------------------------------
| Test cases with children nodes.
|--------------------------------------------------
*/
describe('Modal', () => {
  const closeHandler = jest.fn(() => {});

  it('should redners without crashing when child element present.', () => { 
    const wrapperWithChild = shallow(<Modal closeHandler={closeHandler}>Test</Modal>);
    expect(wrapperWithChild).toMatchSnapshot();
  });
  
  it('should renders without crashing when no child present', () => {
    const wrapper = shallow(<Modal closeHandler={closeHandler} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should not animate', () => {
    const wrapper = mount(<Modal closeHandler={closeHandler} withAnimation={false} />);
    const modalContent = wrapper.find('.modal-content');
    const modalStyle =  modalContent.prop('style');
    expect(modalStyle.marginTop).toBe('0');
    wrapper.unmount();
  })
});

/**
|--------------------------------------------------
| Test Life cyclehooks.
|--------------------------------------------------
*/
describe('Modal behaviour when opened intially', () => {
  let callback = jest.fn(() => {});
  let wrapper = shallow(<Modal closeHandler={callback} isOpen={true}></Modal>);

  it('should render modal without crashing', () => {  
    expect(wrapper).toMatchSnapshot();
  });

  it('should set state correctly', () => {
    expect(wrapper.state('show')).toBe(true);
    expect(wrapper.state('animation')).toBe(false);
  });

  describe('when starts rendering on UI', () => {


    it('should call renderModal Method', () => {
      let spy = jest.spyOn(Modal.prototype, 'renderModal');
      let modal = mount(<Modal closeHandler={callback} isOpen={true}>Test</Modal>);
      expect(spy).toHaveBeenCalled();
      spy.mockRestore();
      modal.unmount();
    });

    it('should set animation after specified delay time', () => {
      jest.useFakeTimers();
      let modal = mount(<Modal closeHandler={callback} isOpen={true}>Test</Modal>);

      expect(setTimeout).toHaveBeenCalled();
      expect(setTimeout.mock.calls[0][1]).toBe(modal.prop('delay'));


      jest.runAllTimers();
      expect(modal.state('show')).toBe(true);
      expect(modal.state('animation')).toBe(true);

      modal.unmount();
      
    });

  });

});

/**
|--------------------------------------------------
| Testing Events.
|--------------------------------------------------
*/

describe('Click Events', () => {
  
  it('should call given callback function on clicking on X icon', () => {
    const callback = jest.fn();
    const modal = mount(<Modal isOpen={true} closeHandler={callback}  title="test"></Modal>);
    expect(callback.mock.calls.length).toBe(0);
    const crossIcon = modal.find('.modal-header > .close > span');
    crossIcon.simulate('click');

    expect(callback).toHaveBeenCalled();
    modal.unmount();
  });

  it('should call given callback on clicking on close button', () => {
    const callback = jest.fn();
    const modal = mount(<Modal isOpen={true} closeHandler={callback}  title="test"></Modal>);

    expect(callback.mock.calls.length).toBe(0);
    const closeButton = modal.find('.modal-footer > button');
    closeButton.simulate('click');

    expect(callback).toHaveBeenCalled();
    modal.unmount();
  });

});
