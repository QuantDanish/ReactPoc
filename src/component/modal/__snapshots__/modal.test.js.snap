// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal behaviour when opened intially should render modal without crashing 1`] = `
<ModalTemplate
  backdropStyle={Object {}}
  body={
    <ModalBody
      hideCallback={[MockFunction]}
    />
  }
  className=""
  hideCallback={[MockFunction]}
  modalStyle={Object {}}
  show={true}
  startAnimation={false}
  withAnimation={true}
/>
`;

exports[`Modal should redners without crashing when child element present. 1`] = `
<ModalTemplate
  backdropStyle={Object {}}
  body="Test"
  className=""
  hideCallback={[MockFunction]}
  modalStyle={Object {}}
  show={false}
  startAnimation={false}
  withAnimation={true}
/>
`;

exports[`Modal should renders without crashing when no child present 1`] = `
<ModalTemplate
  backdropStyle={Object {}}
  body={
    <ModalBody
      hideCallback={[MockFunction]}
    />
  }
  className=""
  hideCallback={[MockFunction]}
  modalStyle={Object {}}
  show={false}
  startAnimation={false}
  withAnimation={true}
/>
`;
